### ISSUE TITLE
TypeError in PngImagePlugin._save

### ISSUE THREAD

**Issue by codewarrior0** [python-pillow/Pillow#1462]
Using Python 2.7.9 for 64-bit Windows. Pillow 3.0.0 installed via `pip install pillow`. Script file:

```
import os

import PIL.Image
basedir = "../data/PIL_Images"
im = PIL.Image.open(os.path.join(basedir, "tinysample.tiff"))
im.save(os.path.join(basedir, "tinysample.png"))
```

Error traceback:
```
Traceback (most recent call last):
  File "pyi_lib_PIL_img_conversion.py", line 17, in <module>
    im.save(os.path.join(basedir, "tinysample.png"))
  File "c:\Users\Rio\Documents\src\pyinst1328\.env27\lib\site-packages\PIL\Image.py", line 1665, in save
    save_handler(self, fp, filename)
  File "c:\Users\Rio\Documents\src\pyinst1328\.env27\lib\site-packages\PIL\PngImagePlugin.py", line 757, in _save
    data = name + b"\0\0" + zlib.compress(im.info["icc_profile"])
TypeError: must be string or read-only buffer, not tuple
```

[Input image file is here.](https://www.dropbox.com/s/4o6zm09was4hf18/tinysample.tiff?dl=0)

This code previously worked with `pillow==2.9.0` installed.

With `pillow==3.0.0`, `im.info["icc_profile"]` is a tuple of length 1 containing a single bytestring. With `pillow==2.9.0`, `im.info["icc_profile"]` is a bytestring.

**Comment by wiredfool** [comment]
Likely fallout from the tiff ifd changes.  Sounds like something is pulling from the legacy system that should be pulling from the _v2 system.

**Comment by wiredfool** [comment]
Git bisect is giving me the TiffIFD rearrangement for where this broke.

**Comment by wiredfool** [comment]
Can you check out the fix here? https://github.com/wiredfool/Pillow/commit/157df44e51ef1436fe3a878b835ec8b4c57d19b0

Also, is it ok to distribute the sample tiff file under the MIT license?

**Comment by codewarrior0** [comment]
Looks like the fix worked. Thanks! 

I don't know the license status of that tiff file. It's part of our test suite, so presumably it's under the same license as the rest of PyInstaller, which is GPL with an exception for the startup code in built applications.

**Comment by tristan** [comment]
perhaps i'm missing something, but doesn't the changes in https://github.com/python-pillow/Pillow/pull/1539 reopen this issue? I think this has been missed since the `test_iccprofile` test case doesn't attempt to save the image as a `PNG`, which is what triggered this issue initially.

e.g. if i change the test case to try and save the tif as a png
```
Pillow tristan$ git diff
diff --git a/Tests/test_file_tiff_metadata.py b/Tests/test_file_tiff_metadata.py
index 1b88fca..cc42163 100644
--- a/Tests/test_file_tiff_metadata.py
+++ b/Tests/test_file_tiff_metadata.py
@@ -172,7 +172,7 @@ class TestFileTiffMetadata(PillowTestCase):
         im = Image.open('Tests/images/hopper.iccprofile.tif')
         out = self.tempfile('temp.tiff')

-        im.save(out)
+        im.save(out, "PNG")
         reloaded = Image.open(out)
         self.assert_(type(im.info['icc_profile']) is not type(tuple))
         self.assertEqual(im.info['icc_profile'], reloaded.info['icc_profile'])
```
the test case now fails with the original error reported in this issue:
```
Pillow tristan$ python Tests/test_file_tiff_metadata.py
..E=== orphaned temp file: /var/folders/f_/mzl0w36s1kvct7sg_zm4n__m0000gn/T/temprzr2wA.tiff
...=== orphaned temp file: /var/folders/f_/mzl0w36s1kvct7sg_zm4n__m0000gn/T/tempggjXpC.tif
.=== orphaned temp file: /var/folders/f_/mzl0w36s1kvct7sg_zm4n__m0000gn/T/tempc2grCt.tiff
.
======================================================================
ERROR: TestFileTiffMetadata.test_iccprofile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "Tests/test_file_tiff_metadata.py", line 175, in test_iccprofile
    im.save(out, "PNG")
  File "build/bdist.macosx-10.10-x86_64/egg/PIL/Image.py", line 1675, in save
    save_handler(self, fp, filename)
  File "build/bdist.macosx-10.10-x86_64/egg/PIL/PngImagePlugin.py", line 757, in _save
    data = name + b"\0\0" + zlib.compress(im.info["icc_profile"])
TypeError: must be string or read-only buffer, not tuple

----------------------------------------------------------------------
Ran 8 tests in 0.033s

FAILED (errors=1)
```

rolling back the changes in https://github.com/python-pillow/Pillow/pull/1539 fixes this, but of course fails for the test case added in https://github.com/python-pillow/Pillow/pull/1539

this is running on OSX, with python 2.7.10 on the current master
```
Pillow tristan$ python -c "import PIL; print PIL.PILLOW_VERSION"
3.2.0.dev0
```

**Comment by znerol** [comment]
I agree with @tristan this issue still exists. Steps to reproduce:

    $ cat <<EOF > test.py
    from __future__ import print_function
    
    import sys
    from PIL import Image, PILLOW_VERSION
    
    img = Image.open(sys.argv[1])
    print(PILLOW_VERSION)
    print(type(img.info['icc_profile']))
    EOF

    $ pip install -q --user pillow==2.9.0 && python test.py ~/Desktop/Trichoglossus_moluccanus_with_open_wings\,_Brisbane-CMYK.tif
    2.9.0
    <type 'str'>

    $ pip install -q --user pillow==3.0.0 && python test.py ~/Desktop/Trichoglossus_moluccanus_with_open_wings\,_Brisbane-CMYK.tif 
    3.0.0
    <type 'tuple'>

    $ pip install -q --user pillow==3.1.0 && python test.py ~/Desktop/Trichoglossus_moluccanus_with_open_wings\,_Brisbane-CMYK.tif 
    3.1.0
    <type 'tuple'>

Test image was converted to CMYK with GIMP as outlined in this [ask ubuntu question](https://askubuntu.com/questions/114858/how-to-convert-image-to-cmyk-in-gimp). The Separate+ plugin is in `gimp-plugin-registry` package on debian/ubuntu. Make sure that the color profile gets actually embedded when exporting the image.

My test image is from [wikipedia](https://commons.wikimedia.org/wiki/File:Trichoglossus_moluccanus_with_open_wings,_Brisbane.jpg), I've attached the CMYK version for your convenience.

[icc profile test images.zip](https://github.com/python-pillow/Pillow/files/170374/icc.profile.test.images.zip)

**Comment by znerol** [comment]
Oh, this is also reproducible using [Tests/images/hopper.iccprofile.tif](../blob/master/Tests/images/hopper.iccprofile.tif)

**Comment by void42** [comment]
This is still reproducible in master using [this TIFF file](https://drive.google.com/file/d/0B0PanjXpbSyHS1l1OU5iRUtYZDQ/view?usp=sharing)

```
image = Image.open(image_file)
image.save('test.png')
```

results in

```
builtins.TypeError
TypeError: a bytes-like object is required, not 'tuple'
```

**Comment by hugovk** [comment]
Reproducible with latest Pillow 3.2.0 (Python 2.7 / OS X):

```python
>>> from PIL import Image
>>> image_file = "LI_MAG_P_20110810_017.tif"
>>> image = Image.open(image_file)
>>> image.save('test.png')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/PIL/Image.py", line 1687, in save
    save_handler(self, fp, filename)
  File "/usr/local/lib/python2.7/site-packages/PIL/PngImagePlugin.py", line 757, in _save
    data = name + b"\0\0" + zlib.compress(im.info["icc_profile"])
TypeError: must be string or read-only buffer, not tuple
```

Pillow 2.9.0 passes, Pillow 3.0.0 errors.

**Comment by jespino** [comment]
I think the problem is setting again the ICCPROFILE length to 0 (this makes the im.info['icc_profile'] to be a tuple), this was in the commit a14467f10bfc269511f4bffcf6b64fbab94ac959 of @wiredfool. The icc_profile haven't a fixed length in the number of bytes, but have a fixed length in the number of values (I'm not sure, but i think so).

Anyway the unique usage of the length in the TiffImagePlugin.py file is for define where to return a value or a tuple. To be compatible with PNG this must be a bytes and not a tuple.

I going to make a small PR with this change to discuss there.

**Comment by wiredfool** [comment]
Length 0 means that it's of indeterminate length. 

Is there a test that can be done with the test files that we have that can demonstrate this? Or do we need to add an image for a useful test case?

**Comment by znerol** [comment]
This is also reproducible using the following file which is already included [Tests/images/hopper.iccprofile.tif](../blob/master/Tests/images/hopper.iccprofile.tif)

**Comment by znerol** [comment]
Oh! There is even a test which is supposed to check that `icc_profil` is not a tuple. The problem is that the assertion is wrong in [Tests/test_file_tiff_metadata.py#L178](../blob/master/Tests/test_file_tiff_metadata.py#L178). `type(tuple)` returns the class `type`. That line should be changed into `self.assert_(type(im.info['icc_profile']) is not tuple)`

It would be better if we'd do a positive assertions, i.e. check the expected type (`str`) instead of rejecting the unexpected (`tuple`).

**Comment by wiredfool** [comment]
That sounds familiar as well.

 #1582. 

 Which has been linked back in to this discussion before. Sigh.

**Comment by znerol** [comment]
So, `git bisect` tells me the following:

The functionality was broken the first time in by 974bcc0 then it was subsequently fixed by 157df44 and then broken again by a14467f

I do not really understand the reason for the last commit. The test added in the second last would have turned red if it would have tested for the right thing but we know it did not.

Can't we just revert a14467f and also fix the test assertion?

**Comment by znerol** [comment]
Okay, seems that does not work since reverting the `length` back to `0` will lead to errors when loading binary profiles.

```
$ make test
python test-installed.py
........SSSSS...............................................................................................................................SSSSSSS....................................................................S...........................S..................................................S...............................................................................................SS.....................SS............................................................S.........S.........SSSSSS.....................................................................................................SSSSSSSSSSSSSSSSSSSSS................SSSSSS.........E....S..........S.......................................
======================================================================
ERROR: TestFileTiffMetadata.test_iccprofile_binary
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/lo/src/Pillow/Tests/test_file_tiff_metadata.py", line 185, in test_iccprofile_binary
    im = Image.open('Tests/images/hopper.iccprofile_binary.tif')
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/Image.py", line 2299, in open
    im = _open_core(fp, filename, prefix)
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/Image.py", line 2289, in _open_core
    im = factory(fp, filename)
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/ImageFile.py", line 97, in __init__
    self._open()
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/TiffImagePlugin.py", line 922, in _open
    self._seek(0)
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/TiffImagePlugin.py", line 983, in _seek
    self._setup()
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/TiffImagePlugin.py", line 1246, in _setup
    a = self._decoder(rawmode, l, i)
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/TiffImagePlugin.py", line 1013, in _decoder
    self.info['icc_profile'] = self.tag_v2[ICCPROFILE]
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/TiffImagePlugin.py", line 500, in __getitem__
    self[tag] = handler(self, data, self.legacy_api)  # check type
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/TiffImagePlugin.py", line 514, in __setitem__
    self._setitem(tag, value, self.legacy_api)
  File "/tmp/pillow-test/lib/python2.7/site-packages/PIL/TiffImagePlugin.py", line 557, in _setitem
    dest[tag], = values
ValueError: too many values to unpack

----------------------------------------------------------------------
Ran 710 tests in 6.309s

FAILED (SKIP=56, errors=1)
Makefile:81: recipe for target 'test' failed
make: *** [test] Error 1
```

**Comment by jespino** [comment]
Yes, I have changed my PR #1831 to process the ICC_PROFILE in a special way that works with PNG and doesn't broke the tests.

**Comment by znerol** [comment]
Oh, did not see PR #1831 maybe that makes more sense. The problem with the current test is that it is not testing the right thing.

**Comment by jespino** [comment]
Yes, I have tested that this works on tiff to png conversion locally,
probably a good thing is add a test to ensure that the conversión between
tiff and png works.

**Comment by znerol** [comment]
@wiredfool is `load_byte` implement correctly in [PIL/TiffImagePlugin.py#L605](../blob/master/PIL/TiffImagePlugin.py#L605)? It seems a bit weird that if `legacy_api` is `True` then it returns same type as `data` but if it is `False` then it simply returns the `data` converted into a `tuple` of the same `length`. As far as I can tell there is no test coverage neither for `load_byte` with `legacy_api=False`.

**Comment by znerol** [comment]
Ok, realized that `load_byte` is covered completely by `test_file_tiff.py`. I regrettably do not figure out how this is supposed to be wired together, so probably cannot help here.

**Comment by wiredfool** [comment]
Yeah. It's complicated. Changing interfaces is something that should be done with great care and more testing than the tiff reorg had.

**Comment by wiredfool** [comment]
Fixed in #1988

**Comment by znerol** [comment]
Repeating the test from one of my [previous comments](#issuecomment-195723952) gives the following correct result:

```
python test.py Trichoglossus_moluccanus_with_open_wings,_Brisbane-CMYK.tif 
3.3.0.dev0
<type 'str'>
```

Thanks a lot.
